File: backlight_sim/sim/sampling.py
"""Ray direction sampling utilities for Monte Carlo simulation."""

from __future__ import annotations

import numpy as np


def sample_isotropic(n: int, rng: np.random.Generator) -> np.ndarray:
    """Sample n uniformly distributed unit vectors on the sphere.

    Returns (n, 3) array.
    """
    z = rng.uniform(-1.0, 1.0, size=n)
    phi = rng.uniform(0.0, 2.0 * np.pi, size=n)
    r = np.sqrt(1.0 - z * z)
    x = r * np.cos(phi)
    y = r * np.sin(phi)
    return np.column_stack([x, y, z])


def sample_lambertian(n: int, normal: np.ndarray, rng: np.random.Generator) -> np.ndarray:
    """Sample n cosine-weighted directions in the hemisphere around normal.

    Uses Malley's method: sample uniform disk, project to hemisphere.
    Returns (n, 3) array of unit vectors.
    """
    normal = np.asarray(normal, dtype=float)
    normal = normal / np.linalg.norm(normal)

    # Build orthonormal basis (tangent, bitangent, normal)
    tangent, bitangent = _build_basis(normal)

    # Malley's method: uniform disk sampling then project up
    r = np.sqrt(rng.uniform(0.0, 1.0, size=n))
    phi = rng.uniform(0.0, 2.0 * np.pi, size=n)
    x_local = r * np.cos(phi)
    y_local = r * np.sin(phi)
    z_local = np.sqrt(np.maximum(0.0, 1.0 - x_local * x_local - y_local * y_local))

    # Transform to world coordinates
    directions = (
        x_local[:, None] * tangent[None, :]
        + y_local[:, None] * bitangent[None, :]
        + z_local[:, None] * normal[None, :]
    )
    return directions


def sample_angular_distribution(
    n: int,
    normal: np.ndarray,
    theta_deg: np.ndarray,
    intensity: np.ndarray,
    rng: np.random.Generator,
) -> np.ndarray:
    """Sample directions from a 1D angular distribution I(theta).

    theta_deg is polar angle in degrees around *normal*.
    """
    theta_deg = np.asarray(theta_deg, dtype=float).reshape(-1)
    intensity = np.asarray(intensity, dtype=float).reshape(-1)
    if theta_deg.size < 2 or intensity.size != theta_deg.size:
        return sample_lambertian(n, normal, rng)

    order = np.argsort(theta_deg)
    theta_deg = theta_deg[order]
    intensity = np.clip(intensity[order], 0.0, None)

    theta_rad = np.radians(theta_deg)
    grid = np.linspace(theta_rad[0], theta_rad[-1], 2048)
    interp_i = np.interp(grid, theta_rad, intensity, left=0.0, right=0.0)
    weights = interp_i * np.sin(grid)
    csum = np.cumsum(weights)
    if csum[-1] <= 0:
        return sample_lambertian(n, normal, rng)
    cdf = csum / csum[-1]

    u = rng.uniform(0.0, 1.0, size=n)
    sample_theta = np.interp(u, cdf, grid)
    phi = rng.uniform(0.0, 2.0 * np.pi, size=n)

    sin_t = np.sin(sample_theta)
    x_local = sin_t * np.cos(phi)
    y_local = sin_t * np.sin(phi)
    z_local = np.cos(sample_theta)

    normal = np.asarray(normal, dtype=float)
    ln = np.linalg.norm(normal)
    if ln <= 0:
        normal = np.array([0.0, 0.0, 1.0], dtype=float)
    else:
        normal = normal / ln
    tangent, bitangent = _build_basis(normal)
    return (
        x_local[:, None] * tangent[None, :]
        + y_local[:, None] * bitangent[None, :]
        + z_local[:, None] * normal[None, :]
    )


def sample_diffuse_reflection(n: int, normal: np.ndarray, rng: np.random.Generator) -> np.ndarray:
    """Sample diffuse (Lambertian) reflection directions off a surface."""
    return sample_lambertian(n, normal, rng)


def reflect_specular(directions: np.ndarray, normal: np.ndarray) -> np.ndarray:
    """Compute specular reflection of direction vectors about a normal.

    directions: (n, 3) incoming directions (pointing toward surface)
    normal: (3,) surface normal
    Returns (n, 3) reflected directions.
    """
    normal = np.asarray(normal, dtype=float)
    normal = normal / np.linalg.norm(normal)
    dot = np.dot(directions, normal)  # (n,)
    return directions - 2.0 * dot[:, None] * normal[None, :]


def scatter_haze(directions: np.ndarray, half_angle_deg: float, rng: np.random.Generator) -> np.ndarray:
    """Perturb each direction within a cone of half_angle_deg.

    Uniform distribution within the cone. Returns (n, 3) unit vectors.
    """
    if half_angle_deg <= 0:
        return directions
    n = len(directions)
    max_theta = np.radians(half_angle_deg)
    # Uniform within cone: cos(theta) in [cos(max_theta), 1]
    cos_min = np.cos(max_theta)
    cos_theta = rng.uniform(cos_min, 1.0, size=n)
    sin_theta = np.sqrt(1.0 - cos_theta * cos_theta)
    phi = rng.uniform(0.0, 2.0 * np.pi, size=n)
    result = np.empty_like(directions)
    for i in range(n):
        d = directions[i]
        ln = np.linalg.norm(d)
        if ln < 1e-12:
            result[i] = d
            continue
        d = d / ln
        t, b = _build_basis(d)
        result[i] = (sin_theta[i] * np.cos(phi[i]) * t +
                      sin_theta[i] * np.sin(phi[i]) * b +
                      cos_theta[i] * d)
    return result


def precompute_bsdf_cdfs(bsdf_profile: dict) -> dict:
    """Pre-build 2D CDF arrays from a BSDF profile for fast importance sampling.

    For each theta_in row, builds a 2048-point CDF over theta_out (sin-weighted).

    Parameters
    ----------
    bsdf_profile : dict
        Profile dict as returned by load_bsdf_csv:
        {"theta_in": [...], "theta_out": [...],
         "refl_intensity": [[...], ...], "trans_intensity": [[...], ...]}

    Returns
    -------
    dict
        {
            "theta_in":  np.ndarray (M,),
            "theta_out_grid": np.ndarray (2048,),  # fine grid for CDF inversion
            "refl_cdf":  np.ndarray (M, 2048),     # CDF per theta_in row
            "trans_cdf": np.ndarray (M, 2048),     # CDF per theta_in row
            "refl_total":  np.ndarray (M,),        # total (un-normalised) weight per row
            "trans_total": np.ndarray (M,),        # total (un-normalised) weight per row
        }
        theta_out_grid is in radians.
    """
    N_GRID = 2048

    theta_in = np.asarray(bsdf_profile["theta_in"], dtype=float)
    theta_out_deg = np.asarray(bsdf_profile["theta_out"], dtype=float)
    refl_mat = np.asarray(bsdf_profile["refl_intensity"], dtype=float)   # (M, N)
    trans_mat = np.asarray(bsdf_profile["trans_intensity"], dtype=float)  # (M, N)

    M = len(theta_in)
    theta_out_rad = np.radians(theta_out_deg)

    # Fine interpolation grid for theta_out
    to_min = theta_out_rad[0] if len(theta_out_rad) > 0 else 0.0
    to_max = theta_out_rad[-1] if len(theta_out_rad) > 0 else np.pi / 2
    grid = np.linspace(to_min, to_max, N_GRID)

    refl_cdf = np.zeros((M, N_GRID), dtype=float)
    trans_cdf = np.zeros((M, N_GRID), dtype=float)
    refl_total = np.zeros(M, dtype=float)
    trans_total = np.zeros(M, dtype=float)

    for i in range(M):
        # Interpolate intensity onto fine grid (sin-weighted for solid angle)
        r_interp = np.interp(grid, theta_out_rad, refl_mat[i], left=0.0, right=0.0)
        t_interp = np.interp(grid, theta_out_rad, trans_mat[i], left=0.0, right=0.0)
        sin_w = np.sin(grid)

        r_weights = np.maximum(r_interp, 0.0) * sin_w
        t_weights = np.maximum(t_interp, 0.0) * sin_w

        r_csum = np.cumsum(r_weights)
        t_csum = np.cumsum(t_weights)

        refl_total[i] = r_csum[-1]
        trans_total[i] = t_csum[-1]

        if refl_total[i] > 0:
            refl_cdf[i] = r_csum / refl_total[i]
        else:
            refl_cdf[i] = np.linspace(0.0, 1.0, N_GRID)  # fallback: uniform

        if trans_total[i] > 0:
            trans_cdf[i] = t_csum / trans_total[i]
        else:
            trans_cdf[i] = np.linspace(0.0, 1.0, N_GRID)  # fallback: uniform

    return {
        "theta_in": theta_in,
        "theta_out_grid": grid,
        "refl_cdf": refl_cdf,
        "trans_cdf": trans_cdf,
        "refl_total": refl_total,
        "trans_total": trans_total,
    }


def sample_bsdf(
    n: int,
    incident_dirs: np.ndarray,
    surface_normal: np.ndarray,
    bsdf_profile: dict,
    mode: str,
    rng: np.random.Generator,
    cdfs: dict | None = None,
) -> np.ndarray:
    """Sample n scattered directions from a 2D BSDF profile.

    Parameters
    ----------
    n : int
        Number of rays to scatter.
    incident_dirs : (n, 3) array
        Incoming ray direction unit vectors (pointing TOWARD the surface).
    surface_normal : (3,) array
        Surface normal (pointing away from the surface on the side the ray arrives from).
    bsdf_profile : dict
        BSDF profile dict (from load_bsdf_csv or precompute_bsdf_cdfs).
    mode : str
        "reflect" — sampled directions in same hemisphere as surface_normal.
        "transmit" — sampled directions in opposite hemisphere.
    rng : np.random.Generator
        Random number generator.
    cdfs : dict, optional
        Pre-computed CDFs from precompute_bsdf_cdfs(). If None, they will be
        computed on-the-fly (less efficient for repeated calls).

    Returns
    -------
    (n, 3) array of unit scattered direction vectors.
    """
    surface_normal = np.asarray(surface_normal, dtype=float)
    ln = np.linalg.norm(surface_normal)
    if ln < 1e-12:
        surface_normal = np.array([0.0, 0.0, 1.0])
    else:
        surface_normal = surface_normal / ln

    incident_dirs = np.asarray(incident_dirs, dtype=float)

    # Build CDFs if not pre-computed
    if cdfs is None:
        cdfs = precompute_bsdf_cdfs(bsdf_profile)

    theta_in_vals = cdfs["theta_in"]           # (M,)
    grid = cdfs["theta_out_grid"]              # (2048,)
    cdf_mat = cdfs["refl_cdf"] if mode == "reflect" else cdfs["trans_cdf"]  # (M, 2048)

    M = len(theta_in_vals)

    # Compute theta_in per ray: angle between -d and surface_normal
    cos_i = np.clip(np.einsum("ij,j->i", -incident_dirs, surface_normal), 0.0, 1.0)
    theta_i_deg = np.degrees(np.arccos(cos_i))  # (n,)

    # Find nearest theta_in bin per ray
    bin_idx = np.searchsorted(theta_in_vals, theta_i_deg, side="right") - 1
    bin_idx = np.clip(bin_idx, 0, M - 1)

    # Sample theta_out via CDF inversion per ray
    u = rng.uniform(0.0, 1.0, size=n)
    sample_theta = np.empty(n, dtype=float)

    # Group rays by theta_in bin for vectorised CDF inversion
    unique_bins = np.unique(bin_idx)
    for b in unique_bins:
        ray_mask = bin_idx == b
        cdf_row = cdf_mat[b]                       # (2048,)
        u_sub = u[ray_mask]
        sample_theta[ray_mask] = np.interp(u_sub, cdf_row, grid)

    # Random azimuth
    phi = rng.uniform(0.0, 2.0 * np.pi, size=n)

    sin_t = np.sin(sample_theta)
    x_local = sin_t * np.cos(phi)
    y_local = sin_t * np.sin(phi)
    z_local = np.cos(sample_theta)

    # Build an outgoing normal for direction hemisphere
    if mode == "transmit":
        hemi_normal = -surface_normal
    else:
        hemi_normal = surface_normal

    tangent, bitangent = _build_basis(hemi_normal)
    directions = (
        x_local[:, None] * tangent[None, :]
        + y_local[:, None] * bitangent[None, :]
        + z_local[:, None] * hemi_normal[None, :]
    )

    # Normalize for safety
    norms = np.linalg.norm(directions, axis=1, keepdims=True)
    norms = np.maximum(norms, 1e-12)
    return directions / norms


def _build_basis(normal: np.ndarray) -> tuple[np.ndarray, np.ndarray]:
    """Build an orthonormal tangent/bitangent pair for a given normal."""
    if abs(normal[0]) < 0.9:
        ref = np.array([1.0, 0.0, 0.0])
    else:
        ref = np.array([0.0, 1.0, 0.0])
    tangent = np.cross(normal, ref)
    tangent = tangent / np.linalg.norm(tangent)
    bitangent = np.cross(normal, tangent)
    return tangent, bitangent

File: backlight_sim/sim/_blu_tracer/src/sampling.cpp
#include "sampling.hpp"
#include <cmath>
#include <algorithm>
#include <vector>

static constexpr double PI = 3.14159265358979323846;

// ---------------------------------------------------------------------------
// Orthonormal basis construction.
// Matches sampling.py::_build_basis: picks ref=(1,0,0) unless |wx| >= 0.9,
// in which case ref=(0,1,0). tangent = cross(w, ref) / |.|; bitangent = cross(w, tangent).
// ---------------------------------------------------------------------------
void build_basis(double wx, double wy, double wz,
                 double& ux, double& uy, double& uz,
                 double& vx, double& vy, double& vz)
{
    double ref_x = 1.0, ref_y = 0.0, ref_z = 0.0;
    if (std::abs(wx) >= 0.9) { ref_x = 0.0; ref_y = 1.0; ref_z = 0.0; }
    // tangent = cross(w, ref)
    ux = wy*ref_z - wz*ref_y;
    uy = wz*ref_x - wx*ref_z;
    uz = wx*ref_y - wy*ref_x;
    double norm_u = std::sqrt(ux*ux + uy*uy + uz*uz);
    if (norm_u < 1e-12) norm_u = 1.0;
    ux /= norm_u; uy /= norm_u; uz /= norm_u;
    // bitangent = cross(w, tangent)
    vx = wy*uz - wz*uy;
    vy = wz*ux - wx*uz;
    vz = wx*uy - wy*ux;
    double norm_v = std::sqrt(vx*vx + vy*vy + vz*vz);
    if (norm_v < 1e-12) norm_v = 1.0;
    vx /= norm_v; vy /= norm_v; vz /= norm_v;
}

void sample_isotropic(int n, std::mt19937_64& rng, RayBatch& batch) {
    std::uniform_real_distribution<double> uni_z(-1.0, 1.0);
    std::uniform_real_distribution<double> uni_phi(0.0, 2.0*PI);
    for (int i = 0; i < n; ++i) {
        const double z = uni_z(rng);
        const double phi = uni_phi(rng);
        const double r = std::sqrt(std::max(0.0, 1.0 - z*z));
        batch.dx[i] = r * std::cos(phi);
        batch.dy[i] = r * std::sin(phi);
        batch.dz[i] = z;
    }
}

void sample_lambertian(int n, double nx, double ny, double nz,
                       std::mt19937_64& rng, RayBatch& batch)
{
    // Normalize the normal defensively
    double nn = std::sqrt(nx*nx + ny*ny + nz*nz);
    if (nn < 1e-12) { nx = 0.0; ny = 0.0; nz = 1.0; nn = 1.0; }
    nx /= nn; ny /= nn; nz /= nn;

    double ux, uy, uz, vx, vy, vz;
    build_basis(nx, ny, nz, ux, uy, uz, vx, vy, vz);
    std::uniform_real_distribution<double> uni(0.0, 1.0);
    for (int i = 0; i < n; ++i) {
        const double r = std::sqrt(uni(rng));
        const double phi = uni(rng) * 2.0*PI;
        const double xl = r * std::cos(phi);
        const double yl = r * std::sin(phi);
        const double zl = std::sqrt(std::max(0.0, 1.0 - xl*xl - yl*yl));
        batch.dx[i] = xl*ux + yl*vx + zl*nx;
        batch.dy[i] = xl*uy + yl*vy + zl*ny;
        batch.dz[i] = xl*uz + yl*vz + zl*nz;
    }
}

void sample_angular_distribution(int n,
                                  double dir_x, double dir_y, double dir_z,
                                  const std::vector<double>& theta_deg,
                                  const std::vector<double>& intensity,
                                  std::mt19937_64& rng, RayBatch& batch)
{
    if (theta_deg.size() < 2 || intensity.size() != theta_deg.size()) {
        sample_lambertian(n, dir_x, dir_y, dir_z, rng, batch);
        return;
    }
    // Normalize direction
    double nn = std::sqrt(dir_x*dir_x + dir_y*dir_y + dir_z*dir_z);
    if (nn < 1e-12) { dir_x = 0.0; dir_y = 0.0; dir_z = 1.0; nn = 1.0; }
    dir_x /= nn; dir_y /= nn; dir_z /= nn;

    // Build 2048-point grid and CDF (matches sampling.py::sample_angular_distribution)
    const int NGRID = 2048;
    const double theta_min_rad = theta_deg.front() * PI / 180.0;
    const double theta_max_rad = theta_deg.back()  * PI / 180.0;
    std::vector<double> grid(NGRID);
    for (int k = 0; k < NGRID; ++k) {
        grid[k] = theta_min_rad + (theta_max_rad - theta_min_rad) * k / (NGRID - 1);
    }

    auto interp_deg = [&](double x_rad) -> double {
        const double x_deg = x_rad * 180.0 / PI;
        if (x_deg <= theta_deg.front()) return intensity.front();
        if (x_deg >= theta_deg.back())  return intensity.back();
        auto it = std::lower_bound(theta_deg.begin(), theta_deg.end(), x_deg);
        const int hi = (int)(it - theta_deg.begin());
        const int lo = hi - 1;
        const double denom = theta_deg[hi] - theta_deg[lo];
        const double frac = denom > 0.0 ? (x_deg - theta_deg[lo]) / denom : 0.0;
        return intensity[lo] + frac*(intensity[hi] - intensity[lo]);
    };

    std::vector<double> cdf(NGRID);
    double total = 0.0;
    for (int k = 0; k < NGRID; ++k) {
        const double w = std::max(0.0, interp_deg(grid[k])) * std::sin(grid[k]);
        total += w;
        cdf[k] = total;
    }
    if (total <= 0.0) {
        sample_lambertian(n, dir_x, dir_y, dir_z, rng, batch);
        return;
    }
    for (auto& c : cdf) c /= total;

    double ux, uy, uz, vx, vy, vz;
    build_basis(dir_x, dir_y, dir_z, ux, uy, uz, vx, vy, vz);
    std::uniform_real_distribution<double> uni(0.0, 1.0);
    for (int i = 0; i < n; ++i) {
        const double u_val = uni(rng);
        auto it = std::lower_bound(cdf.begin(), cdf.end(), u_val);
        int idx = (int)(it - cdf.begin());
        if (idx >= NGRID) idx = NGRID - 1;
        const double sample_theta = grid[idx];
        const double phi = uni(rng) * 2.0*PI;
        const double sin_t = std::sin(sample_theta);
        const double cos_t = std::cos(sample_theta);
        const double xl = sin_t * std::cos(phi);
        const double yl = sin_t * std::sin(phi);
        const double zl = cos_t;
        batch.dx[i] = xl*ux + yl*vx + zl*dir_x;
        batch.dy[i] = xl*uy + yl*vy + zl*dir_y;
        batch.dz[i] = xl*uz + yl*vz + zl*dir_z;
    }
}

void sample_lambertian_single(double nx, double ny, double nz,
                               std::mt19937_64& rng,
                               double& out_x, double& out_y, double& out_z)
{
    double nn = std::sqrt(nx*nx + ny*ny + nz*nz);
    if (nn < 1e-12) { nx = 0.0; ny = 0.0; nz = 1.0; nn = 1.0; }
    nx /= nn; ny /= nn; nz /= nn;

    double ux, uy, uz, vx, vy, vz;
    build_basis(nx, ny, nz, ux, uy, uz, vx, vy, vz);
    std::uniform_real_distribution<double> uni(0.0, 1.0);
    const double r = std::sqrt(uni(rng));
    const double phi = uni(rng) * 2.0*PI;
    const double xl = r * std::cos(phi);
    const double yl = r * std::sin(phi);
    const double zl = std::sqrt(std::max(0.0, 1.0 - xl*xl - yl*yl));
    out_x = xl*ux + yl*vx + zl*nx;
    out_y = xl*uy + yl*vy + zl*ny;
    out_z = xl*uz + yl*vz + zl*nz;
}

void reflect_specular(double dx, double dy, double dz,
                      double nx, double ny, double nz,
                      double& rdx, double& rdy, double& rdz)
{
    const double dot = dx*nx + dy*ny + dz*nz;
    rdx = dx - 2.0*dot*nx;
    rdy = dy - 2.0*dot*ny;
    rdz = dz - 2.0*dot*nz;
    const double norm = std::sqrt(rdx*rdx + rdy*rdy + rdz*rdz);
    if (norm > 1e-12) { rdx /= norm; rdy /= norm; rdz /= norm; }
}

void scatter_haze_single(double dx, double dy, double dz,
                          double half_angle_deg,
                          std::mt19937_64& rng,
                          double& out_x, double& out_y, double& out_z)
{
    if (half_angle_deg <= 0.0) {
        out_x = dx; out_y = dy; out_z = dz;
        return;
    }
    double nn = std::sqrt(dx*dx + dy*dy + dz*dz);
    if (nn < 1e-12) { out_x = dx; out_y = dy; out_z = dz; return; }
    dx /= nn; dy /= nn; dz /= nn;

    const double half_rad = half_angle_deg * PI / 180.0;
    const double cos_max = std::cos(half_rad);
    std::uniform_real_distribution<double> uni(0.0, 1.0);
    const double cos_t = cos_max + (1.0 - cos_max)*uni(rng);
    const double sin_t = std::sqrt(std::max(0.0, 1.0 - cos_t*cos_t));
    const double phi = uni(rng) * 2.0*PI;
    double ux, uy, uz, vx, vy, vz;
    build_basis(dx, dy, dz, ux, uy, uz, vx, vy, vz);
    out_x = sin_t*std::cos(phi)*ux + sin_t*std::sin(phi)*vx + cos_t*dx;
    out_y = sin_t*std::cos(phi)*uy + sin_t*std::sin(phi)*vy + cos_t*dy;
    out_z = sin_t*std::cos(phi)*uz + sin_t*std::sin(phi)*vz + cos_t*dz;
    const double norm = std::sqrt(out_x*out_x + out_y*out_y + out_z*out_z);
    if (norm > 1e-12) { out_x /= norm; out_y /= norm; out_z /= norm; }
}
